lms/djangoapps/certificates/tests/test_webview_views.py

```diff
@@ -945,6 +964,54 @@ def test_html_view_certificate_available_date_for_instructor_paced_courses(self,
         )
         self.assertContains(response, date)
 
+    @override_settings(FEATURES=FEATURES_WITH_CERTS_ENABLED)
+    @ddt.data(
+        (True, False),
+        (False, False),
+        (True, True),
+        (False, True)
+    )
+    @ddt.unpack
+    def test_html_view_certificate_display_date(self, self_paced, date_override):
+        """
+        Test certificate web view should display the correct date on the
+        certificate in all cases:
+            * self-paced, no date override
+            * instructor-paced with certificate_available_date
+            * self-paced with date override
+            * instructor-paced with date override
+        """
+        self.course.self_paced = self_paced
+        if date_override:
+            self._add_certificate_date_override()
+        today = datetime.datetime.utcnow()
+        self.course.certificate_available_date = today + datetime.timedelta(-2)
+        self.store.update_item(self.course, self.user.id)
+        self._add_course_certificates(count=1, signatory_count=1, is_active=True)
+        test_url = get_certificate_url(
+            user_id=self.user.id,
+            course_id=str(self.course.id),
+            uuid=self.cert.verify_uuid
+        )
+
+        with override_waffle_switch(AUTO_CERTIFICATE_GENERATION, active=True):
+            response = self.client.get(test_url)
+
+        if date_override:
+            expected_date = self.cert.date_override.date
+        elif self_paced or self.course.certificate_available_date > today:
```

Comment:
If it's instructor paced and the CAD is in the future, shouldn't the response be that you can't see the cert (instead of showing it with a date of today)? I might be reading this wrong.

Reply:
Weirdly, this testing logic is correct according to the code. See the existing (and unchanged in my PR) code here:

https://github.com/edx/edx-platform/blob/7d029f82833be5657faa98be305b059d3f9bc0a1/lms/djangoapps/certificates/api.py#L855-L857

This function, `display_date_for_certificate` (being tested here), actually returns the `modified` date if the `certificate_available_date` is in the future. _Why?_ I’m not sure. But this is the existing logic, and messing with it seemed to me to be out of scope for this task. 

(Also, remember: this function is _only about what date is displayed on the certificate_, and **not** about whether or not the certificate is visible.)